pkg/util/tracing/collector/collector.go

```diff
@@ -53,59 +50,125 @@ func New(
+	// recordedSpans, then we can return them when Next() is called.
+	if i.recordedSpans != nil && i.recordedSpanIndex < len(i.recordedSpans) {
+		return true
 	}
 
-	// Collect spans from all clients.
-	// We'll want to rate limit outgoing RPCs (limit pulled out of thin air).
-	var mu syncutil.Mutex
-	qp := quotapool.NewIntPool("every-node", 25)
-	log.Infof(ctx, "executing GetSpanRecordings on nodes %s", ns)
-	grp := ctxgroup.WithContext(ctx)
-
-	for _, node := range ns {
-		id := node.id // copy out of the loop variable
-		alloc, err := qp.Acquire(ctx, 1)
-		if err != nil {
-			return res, err
+	// Reset buffer variables.
+	i.recordedSpans = nil
+	i.recordedSpanIndex = 0
+
+	// Otherwise, either there are no more spans or we have exhausted the
+	// recordings from the current node, and we need to pull the inflight
+	// recordings from another node.
+	// Keep searching for recordings from all live nodes in the cluster.
+	for i.recordedSpans == nil {
+		i.curNodeIndex++
+		// No more spans to return from any of the live nodes in the cluster.
+		if !(i.curNodeIndex < len(i.liveNodes)) {
+			return false
+		}
+		i.recordedSpans, i.iterErr = i.collector.getTraceSpanRecordingsForNode(i.ctx, i.traceID,
+			i.liveNodes[i.curNodeIndex])
+		// TODO(adityamaru): We might want to consider not failing if a single node
+		// fails to return span recordings.
+		if i.iterErr != nil {
+			return false
 		}
+	}
+
+	return true
+}
+
+// Next sets the iterator index to point to the next buffered RecordedSpan.
+func (i *TraceCollectorIterator) Next() {
+	i.recordedSpanIndex++
+}
+
+// Value returns the RecordedSpan pointed to by the iterator.
+func (i *TraceCollectorIterator) Value() tracingpb.RecordedSpan {
+	return i.recordedSpans[i.recordedSpanIndex]
+}
+
+// Error returns the error encountered by the iterator when fetching inflight
+// span recordings.
+func (i *TraceCollectorIterator) Error() error {
+	return i.iterErr
+}
 
-		grp.GoCtx(func(ctx context.Context) error {
-			defer alloc.Release()
-
-			conn, err := t.dialer.Dial(ctx, id, rpc.DefaultClass)
-			if err != nil {
-				return err
-			}
-			traceClient := tracingservicepb.NewTracingClient(conn)
-			resp, err := traceClient.GetSpanRecordings(ctx,
-				&tracingservicepb.SpanRecordingRequest{TraceID: traceID})
-			if err != nil {
-				return err
-			}
-			mu.Lock()
-			res = append(res, resp.SpanRecordings...)
-			mu.Unlock()
-			return nil
-		})
+// getTraceSpanRecordingsForNode returns the inflight span recordings for traces
+// with traceID from the node with nodeID. The span recordings are sorted by
+// StartTime.
+// This method does not distinguish between requests for local and remote
+// inflight spans, and relies on gRPC short circuiting local requests.
+func (t *TraceCollector) getTraceSpanRecordingsForNode(
+	ctx context.Context, traceID uint64, nodeID roachpb.NodeID,
+) ([]tracingpb.RecordedSpan, error) {
+	log.Infof(ctx, "executing GetSpanRecordings on nodes %s", nodeID.String())
```

Comment:
Update this log? Singular form, and can drop the method name (unnecessary, we already know which line it came from).

Reply:
done.